File: java/structures/LinkedLists/SinglyLinkedList.java
package structures.LinkedLists;

public class SinglyLinkedList<T> implements Iterable<T> {

  private int size = 0;
  private Node<T> head = null;

  private static class Node<T> {
    T data;
    Node<T> next;

    Node(T data, Node<T> next) {
      this.data = data;
      this.next = next;
    }
  }

  public void clear() {
    head = null;
    size = 0;
  }

  public int size() {
    return size;
  }

  public boolean isEmpty() {
    return size == 0;
  }

  public void add(T value) {
    addLast(value);
  }

  public void addFirst(T value) {
    head = new Node<>(value, head);
    size++;
  }

  public void addLast(T value) {
    Node<T> newNode = new Node<>(value, null);
    if (isEmpty()) {
      head = newNode;
    } else {
      Node<T> trav = head;
      while (trav.next != null)
        trav = trav.next;
      trav.next = newNode;
    }
    size++;
  }

  public void addAt(int index, T value) {
    if (index < 0 || index > size)
      throw new IndexOutOfBoundsException("Index " + index + " out of bounds");
    if (index == 0) {
      addFirst(value);
      return;
    }
    Node<T> prev = head;
    for (int i = 1; i < index; i++)
      prev = prev.next;
    prev.next = new Node<>(value, prev.next);
    size++;
  }

  public T getFirst() {
    if (isEmpty())
      throw new IllegalStateException("List is empty");
    return head.data;
  }

  public T getLast() {
    if (isEmpty())
      throw new IllegalStateException("List is empty");
    Node<T> trav = head;
    while (trav.next != null)
      trav = trav.next;
    return trav.data;
  }

  public T removeFirst() {
    if (isEmpty())
      throw new IllegalStateException("List is empty");
    T data = head.data;
    head = head.next;
    size--;
    return data;
  }

  public T removeLast() {
    if (isEmpty())
      throw new IllegalStateException("List is empty");
    if (size == 1)
      return removeFirst();
    Node<T> prev = head;
    while (prev.next.next != null)
      prev = prev.next;
    T data = prev.next.data;
    prev.next = null;
    size--;
    return data;
  }

  public T removeAt(int index) {
    if (index < 0 || index >= size)
      throw new IndexOutOfBoundsException("Index " + index + " out of bounds");
    if (index == 0)
      return removeFirst();
    Node<T> prev = head;
    for (int i = 1; i < index; i++)
      prev = prev.next;
    T data = prev.next.data;
    prev.next = prev.next.next;
    size--;
    return data;
  }

  public boolean remove(Object obj) {
    if (isEmpty())
      return false;
    if (obj == null) {
      if (head.data == null) {
        removeFirst();
        return true;
      }
      Node<T> prev = head;
      while (prev.next != null) {
        if (prev.next.data == null) {
          prev.next = prev.next.next;
          size--;
          return true;
        }
        prev = prev.next;
      }
    } else {
      if (obj.equals(head.data)) {
        removeFirst();
        return true;
      }
      Node<T> prev = head;
      while (prev.next != null) {
        if (obj.equals(prev.next.data)) {
          prev.next = prev.next.next;
          size--;
          return true;
        }
        prev = prev.next;
      }
    }
    return false;
  }

  public int indexOf(Object obj) {
    Node<T> trav = head;
    int index = 0;
    if (obj == null) {
      while (trav != null) {
        if (trav.data == null)
          return index;
        trav = trav.next;
        index++;
      }
    } else {
      while (trav != null) {
        if (obj.equals(trav.data))
          return index;
        trav = trav.next;
        index++;
      }
    }
    return -1;
  }

  public boolean contains(Object obj) {
    return indexOf(obj) != -1;
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder("[");
    Node<T> trav = head;
    while (trav != null) {
      sb.append(trav.data);
      if (trav.next != null)
        sb.append(", ");
      trav = trav.next;
    }
    sb.append("]");
    return sb.toString();
  }

  @Override
  public java.util.Iterator<T> iterator() {
    return new java.util.Iterator<T>() {
      private Node<T> current = head;

      @Override
      public boolean hasNext() {
        return current != null;
      }

      @Override
      public T next() {
        if (!hasNext())
          throw new java.util.NoSuchElementException();
        T data = current.data;
        current = current.next;
        return data;
      }
    };
  }

  // -------------- MAIN --------------
  public static void main(String[] args) {
    SinglyLinkedList<Integer> list = new SinglyLinkedList<>();
    System.out.println("Initial: " + list); // Print the initial (empty) list

    // Add an element at the beginning
    list.addFirst(10);

    // Add an element at the end
    list.addLast(20);

    // Add an element at the end (same as addLast)
    list.add(30);

    // Add an element at index 1 (between 10 and 20)
    list.addAt(1, 15);
    System.out.println("After adds: " + list); // Expected: [10, 15, 20, 30]

    // Get first and last elements
    System.out.println("First: " + list.getFirst() + ", Last: " + list.getLast()); // 10, 30

    // Remove the element at index 2 (value 20)
    System.out.println("removeAt(2): " + list.removeAt(2));
    System.out.println("After removeAt: " + list); // Expected: [10, 15, 30]

    // Remove the element with value 15
    list.remove((Integer) 15);
    System.out.println("After remove(15): " + list); // Expected: [10, 30]

    // Iterate through the list
    System.out.print("Iterating: ");
    for (int v : list)
      System.out.print(v + " ");
    System.out.println(); // Output: 10 30

    // Clear the list
    list.clear();
    System.out.println("After clear, isEmpty: " + list.isEmpty()); // true
  }
}

File: python/LinkedLists/SinglyLinkedList.py
from typing import Generic, TypeVar, Optional, Iterator

T = TypeVar('T')


class SinglyLinkedList(Generic[T]):
    class _Node:
        __slots__ = ('data', 'next')

        def __init__(self, data: T, next: Optional['SinglyLinkedList._Node']):
            self.data = data
            self.next = next

        def __str__(self) -> str:
            return str(self.data)

    def __init__(self):
        self._size: int = 0
        self._head: Optional[SinglyLinkedList._Node] = None

    def clear(self) -> None:
        trav = self._head
        while trav is not None:
            nxt = trav.next
            trav.next = None
            trav.data = None  # type: ignore
            trav = nxt
        self._head = None
        self._size = 0

    def size(self) -> int:
        return self._size

    def is_empty(self) -> bool:
        return self._size == 0

    def add(self, value: T) -> None:
        self.add_last(value)

    def add_first(self, value: T) -> None:
        new_node = SinglyLinkedList._Node(value, self._head)
        self._head = new_node
        self._size += 1

    def add_last(self, value: T) -> None:
        new_node = SinglyLinkedList._Node(value, None)
        if self.is_empty():
            self._head = new_node
        else:
            trav = self._head
            while trav.next is not None:
                trav = trav.next
            trav.next = new_node
        self._size += 1

    def add_at(self, index: int, value: T) -> None:
        if index < 0 or index > self._size:
            raise IndexError(f"Index {index} out of bounds")
        if index == 0:
            self.add_first(value)
            return
        prev = self._head
        for _ in range(1, index):
            prev = prev.next  # type: ignore
        node = SinglyLinkedList._Node(value, prev.next)  # type: ignore
        prev.next = node  # type: ignore
        self._size += 1

    def get_first(self) -> T:
        if self.is_empty():
            raise IndexError("List is empty")
        return self._head.data  # type: ignore

    def get_last(self) -> T:
        if self.is_empty():
            raise IndexError("List is empty")
        trav = self._head
        while trav.next is not None:  # type: ignore
            trav = trav.next  # type: ignore
        return trav.data  # type: ignore

    def remove_first(self) -> T:
        if self.is_empty():
            raise IndexError("List is empty")
        data = self._head.data  # type: ignore
        self._head = self._head.next  # type: ignore
        self._size -= 1
        return data

    def remove_last(self) -> T:
        if self.is_empty():
            raise IndexError("List is empty")
        if self._size == 1:
            return self.remove_first()
        prev = self._head
        while prev.next.next is not None:  # type: ignore
            prev = prev.next  # type: ignore
        data = prev.next.data  # type: ignore
        prev.next = None  # type: ignore
        self._size -= 1
        return data

    def remove_at(self, index: int) -> T:
        if index < 0 or index >= self._size:
            raise IndexError(f"Index {index} out of bounds")
        if index == 0:
            return self.remove_first()
        prev = self._head
        for _ in range(1, index):
            prev = prev.next  # type: ignore
        data = prev.next.data  # type: ignore
        prev.next = prev.next.next  # type: ignore
        self._size -= 1
        return data

    def remove(self, value: T) -> bool:
        if self.is_empty():
            return False
        if value is None:
            if self._head.data is None:  # type: ignore
                self.remove_first()
                return True
            prev = self._head
            while prev.next is not None:
                if prev.next.data is None:  # type: ignore
                    prev.next = prev.next.next
                    self._size -= 1
                    return True
                prev = prev.next  # type: ignore
        else:
            if value == self._head.data:  # type: ignore
                self.remove_first()
                return True
            prev = self._head
            while prev.next is not None:
                if value == prev.next.data:  # type: ignore
                    prev.next = prev.next.next
                    self._size -= 1
                    return True
                prev = prev.next  # type: ignore
        return False

    def index_of(self, value: T) -> int:
        trav = self._head
        index = 0
        if value is None:
            while trav is not None:
                if trav.data is None:
                    return index
                trav = trav.next
                index += 1
        else:
            while trav is not None:
                if value == trav.data:
                    return index
                trav = trav.next
                index += 1
        return -1

    def contains(self, value: T) -> bool:
        return self.index_of(value) != -1

    def __iter__(self) -> Iterator[T]:
        trav = self._head
        while trav is not None:
            yield trav.data  # type: ignore
            trav = trav.next  # type: ignore

    def __str__(self) -> str:
        elements = []
        trav = self._head
        while trav is not None:
            elements.append(str(trav.data))
            trav = trav.next
        return f"[{', '.join(elements)}]"


if __name__ == "__main__":
    lst = SinglyLinkedList[int]()
    print("Initial:", lst)  # []

    # Add an element at the beginning
    lst.add_first(10)

    # Add an element at the end
    lst.add_last(20)

    # Add an element at the end (same as add)
    lst.add(30)

    # Add an element at index 1 (between 10 and 20)
    lst.add_at(1, 15)
    print("After adds:", lst)  # [10, 15, 20, 30]

    # Get first and last elements
    # First: 10, Last: 30
    print(f"First: {lst.get_first()}, Last: {lst.get_last()}")

    # Remove the element at index 2 (value 20)
    print("remove_at(2):", lst.remove_at(2))
    print("After remove_at:", lst)  # [10, 15, 30]

    # Remove the element with value 15
    lst.remove(15)
    print("After remove(15):", lst)  # [10, 30]

    # Iterate through the list
    print("Iterating:", end=" ")
    for v in lst:
        print(v, end=" ")
    print()

    # Clear the list
    lst.clear()
    print("After clear, is_empty():", lst.is_empty())  # True
